File: 1-arrays/cpp/189-rotate-array.cpp
#include <bits/stdc++.h>
using namespace std;
void reverse(vector<int>&, int, int);
void rotate(vector<int>&, int);

void rotate(vector<int>& nums, int k) {
    k = k % nums.size();

    reverse(nums, 0, nums.size()-1);
    reverse(nums, 0, k-1);
    reverse(nums, k, nums.size()-1);
}

void reverse(vector<int>& nums, int start, int end){
    while(start < end){
        swap(nums[start++], nums[end--]);

        /* int tmp = nums[start];
        nums[start] = nums[end];
        nums[end] = tmp;
        start++;
        end--; */
    }
}

int main(){
    vector<int> nums = {1,2,3,4,5,6,7};
    vector<int> ans;
    rotate(nums, 3);

    for(int num: nums){
        cout<< num <<" ";
    }
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)

File: 1-arrays/java/189-rotate-array.java
public class Solution{
    public static void rotate(int[] nums, int k){
        k = k % nums.length;

        reverse(nums, 0, nums.length-1);
        reverse(nums, 0, k-1);
        reverse(nums, k, nums.length-1);
    }

    public static void reverse(int[] nums, int start, int end){
        while(start < end){
            int tmp = nums[start];
            nums[start] = nums[end];
            nums[end] = tmp;
            start++;
            end--;
        }
    }
    public static void main(String args[]){
        int nums[] = {1,2,3,4,5,6,7};
        rotate(nums, 3);

        for(int num: nums){
            System.out.print(num + " ");
        }
    }
}

// Time Complexity: O(n)
// Space Complexity: O(1)
